newIDE/app/src/UI/TreeView/index.js

```diff
@@ -0,0 +1,609 @@
+    DragSourceAndDropTarget,
+    getItemHtmlId
+  );
+
+  // Reset opened nodes during search when user stops searching
+  // or when the search text changes.
+  React.useEffect(
+    () => {
+      if (!searchText || searchText.length > 0) {
+        setOpenedDuringSearchNodeIds([]);
+      }
+    },
+    [searchText]
+  );
+
+  const onKeyDown = React.useCallback(
+    (event: KeyboardEvent) => {
+      if (!navigationKeys.includes(event.key)) return;
+      let newFocusedItem;
+      const item = selectedItems[0];
+      let itemIndexInFlattenedData = -1;
+      if (item) {
+        itemIndexInFlattenedData = flattenedData.findIndex(
+          node => node.id === getItemId(item)
+        );
+      }
+
+      if (itemIndexInFlattenedData === -1) {
+        // If no row is selected, start from the first row that is selectable.
+        let i = 0;
+        let newFocusedNode = flattenedData[i];
+        console.log(newFocusedNode);
+        while (
+          newFocusedNode &&
+          (newFocusedNode.item.isRoot || newFocusedNode.item.isPlaceholder)
+        ) {
+          i += 1;
+          if (i > flattenedData.length - 1) {
+            newFocusedNode = null;
+          }
+          newFocusedNode = flattenedData[i];
+        }
+        if (newFocusedNode) {
+          newFocusedItem = newFocusedNode.item;
+        }
+      } else if (event.key === 'ArrowDown') {
+        event.preventDefault();
+        if (itemIndexInFlattenedData < flattenedData.length - 1) {
+          let delta = 1;
+          let newFocusedNode = flattenedData[itemIndexInFlattenedData + delta];
+          while (
+            newFocusedNode &&
+            (newFocusedNode.item.isRoot || newFocusedNode.item.isPlaceholder)
+          ) {
+            if (itemIndexInFlattenedData + delta > flattenedData.length - 1) {
+              newFocusedNode = null;
+            }
+            delta += 1;
+            newFocusedNode = flattenedData[itemIndexInFlattenedData + delta];
+          }
+          if (newFocusedNode) {
+            newFocusedItem = newFocusedNode.item;
+          }
+        }
+      } else if (event.key === 'ArrowUp') {
+        event.preventDefault();
+        if (itemIndexInFlattenedData > 0) {
+          let delta = -1;
+          let newFocusedNode = flattenedData[itemIndexInFlattenedData + delta];
+          while (
+            newFocusedNode &&
+            (newFocusedNode.item.isRoot || newFocusedNode.item.isPlaceholder)
+          ) {
+            if (itemIndexInFlattenedData + delta < 0) {
+              newFocusedNode = null;
+            }
+            delta -= 1;
+            newFocusedNode = flattenedData[itemIndexInFlattenedData + delta];
+          }
+          if (newFocusedNode) {
+            newFocusedItem = newFocusedNode.item;
+          }
+        }
+      } else if (event.key === 'ArrowRight' && arrowKeyNavigationProps) {
+        event.preventDefault();
+        newFocusedItem = arrowKeyNavigationProps.onArrowRight(
```

Comment:
I wonder if we could give a tiny bit less responsibility to the parent:

onArrowRight can either:
- return another item to focus (which is what it does now)
- or return null. In this case, we `closeItems`/`openItems` according to if it's already opened or not. 

I just find it a bit weird that the parent ObjectsList does `treeViewRef.current.openItems([getTreeViewItemId(item)]);` while it could be handled by the tree... the tree needs just to know if there is a children or parent to navigate to, otherwise it can handle closing/opening